usbreset.1.in
.\" SPDX-License-Identifier: GPL-2.0-only
.\" Copyright (c) 2023 Aurelien Jarno <aurelien@aurel32.net>
.TH usbreset 1 "04 January 2024" "usbutils" "Linux USB Utilities"
.IX usbreset
.SH NAME
usbreset \- send a USB port reset to a USB device
.SH SYNOPSIS
.B usbreset
[
.I device
]

.SH DESCRIPTION
.B usbreset
is a utility that performs resets on USB devices. It is particularly useful
situations where a USB device is unresponsive or exhibits erratic behavior.
The USB \fIdevice\fP to be reset can be specified in one of three formats:
.TP
.BR PPPP:VVVV
Reset by product and vendor IDs
.TP
.BR BBB/DDD
Reset by bus and device number
.TP
.BR "Product"
Reset by product name
.PP
When run without any arguments,
.B usbreset
provides usage information and a list of connected USB devices, including their
product and vendor IDs, bus and device numbers, and product names.

.SH RETURN VALUE
If the specified device is not found, a non-zero exit code is returned.

.SH EXAMPLES
.TP
Reset device with vendor ID 1234 and product ID 5678:
.B usbreset 1234:5678

.TP
Reset device 002 on bus 001:
.B usbreset 001:002

.TP
Reset device named USB2.0 Hub:
.B usbreset """USB2.0 Hub"""

.SH SEE ALSO
.BR lsusb (8).

.SH AUTHOR
Alan Stern <stern@rowland.harvard.edu>


NEWS
# usbutils old-style "NEWS" file
# SPDX-License-Identifier: GPL-2.0-only
#
# Copyright (c) 2018-2023 Greg Kroah-Hartman <gregkh@linuxfoundation.org>
#
# This file is here for people who do not have access to the git commit
# history (i.e. packaged versions of the software).
#

usbutils 017
============

Greg Kroah-Hartman (4):
      lsusb: fix up [unknown] vendor and product strings.
      lsusb: fix build warning for dump_billboard_alt_mode_capability_desc()
      lsusb: add fallback names for 'lsusb -v' output
      names: simplify get_vendor_product_with_fallback() a bit

rezso (1):
      Honor system libdir and includedir

usbutils 016
============

Alan Stern (1):
      usbutils: lsusb-t: print entries for devices with no interfaces

Aurelien Jarno (1):
      Fix a typo in usb-spec.h

Bob Hyman (1):
      lsusb.py.in: Display (device) power/wakeup via -w option.

Dingyan Li (3):
      Fix an incorrect length value in hid descriptor.
      Fix misalignments in hid device descripptor.
      Use bigger buffer to place speed value string

Greg Kroah-Hartman (19):
      lsusb -h returns an error
      lsusb -h fixups
      lsusb -t: sort in bus order, not reverse order
      lsusb -t: print ports and busses and devices with same width
      lsusb -t: assign_interface_to_parent() fixups
      lsusb.8.in: fix up missing '-' in text
      README.md: add source location
      lsusb.py: fix up wakeup logic for devices that do not support it
      lsusb.py.in: add another default path for usb.ids
      names.c: if a string can not be found in the usb.ids file, return [unknown]
      lsusb-t: if a driver is not bound to an interface, report "[none]"

Hannes Weisbach (1):
      Generate usbutils.pc pkgconfig file

Kristoffer Ellersgaard Koch (1):
      usbreset: Allow idProduct and idVendor to be 0

Lukas Nykryn (1):
      usb-devices: make shellcheck happy

Tan Li Boon (3):
      lsusb: Add function that sorts the output by device ID.
      lsusb: Additional sorting by bus number.
      lsusb: This is a more compact implementation of the device list sort implemented within this pull request. The output remains the same as the one demonstrated in the previous commit.

usbutils 015
============

Greg Kroah-Hartman (6):
      usb-devices: list the root devices in numerical order
      usb-devices: use 'local' variable type to handle recursion
      lsusb: remove unused wireless check
      lsusb: remove wireless descriptor information
      usb-devices: fix field width on device speed field
      lsusb: fix up Midi Device specification devices

Han Han (1):
      Fix an runtime error reported by undefind sanitizer

Robert Hancock (1):
      lsusb: Improve status display for SuperSpeedPlus hubs

Tan Li Boon (1):
      lsusb-t: Fix recursive sorting on child devices.


usbutils 014
============

Adrian Kalazi (1):
      Add Tx+Rx lanes to tree mode

Aurelien Jarno (1):
      lsusb(8): mention the udev's hardware database

Benson Leung (1):
      lsusb: Fix typo in string output for iAdditionalInfoURL

Bui Quang Minh (1):
      lsusb-t: Add lower bound checks in read_sysfs_file_string

Grant Grundler (2):
      lsusb: remove unused RETRIES constant
      lsusb: don't complain on EAGAIN

Greg Kroah-Hartman (10):
      sysfs: add copyright notice taken from name.*
      LICENSES: add symlink to handle "or-later" issue for GPL-2.0
      LICENSES: put spdx headers on the license files
      rename "GPL-2.0+" to "GPL-2.0-or-later"
      LICENSE: rename GPL-3.0.txt -> GPL-3.0-only.txt
      usbhid-dump.8.in: add copyright information
      .gitignore: add copyright and SPDX info
      usbhid-dump: add copyright and SPDX info
      SPDX header cleanups from GPL-2.0 -> GPL-2.0-only
      usbutils.spdx: update with output of latest reuse tool

Jonathan Neuschäfer (2):
      lsusb: Fix spelling of bEndpointAddress in UVC
      lsusb: Decode endpoint addresses in UVC

Lukas Zaoral (3):
      lsusb.c: fix leak in dump_printer_device
      usb-devices: do not use `local` in a POSIX shell script
      desc-defs.c: fix possible out-of-bound read

Matthias Braun (1):
      Fix typos in lsusb.8.in

Ruslan Kabatsayev (5):
      Fix locating endpoint when it's a directory rather than a symlink
      Fix formatting of interface descriptors to match /sys/kernel/debug/usb/devices
      Fix formatting of endpoint direction to match /sys/kernel/debug/usb/devices
      Fix formatting of endpoint type to match /sys/kernel/debug/usb/devices
      Fix formatting of max endpoint packet size to match /sys/kernel/debug/usb/devices

Thomas Hebb (1):
      lsusb: Fix buffer size copy/paste error

Tormod Volden (1):
      Makefile.am: Remove unused DATADIR


usbutils 013
============

Aurelien Jarno (1):
      lsusb.8.in: do not mention usb.ids

Baruch Siach (1):
      usb-devices.1: don't mention bash

Greg Kroah-Hartman (15):
      usbhid-dump: move manpage to main directory out of subdir
      usbhid-dump.8: add SPDX header
      usbhid-dump.8: autogenerate it with the usbutils version number
      .gitignore: add usbhid-dump.8
      usbhid-dump: add SPDX identifiers to all files.
      usbhid-dump: remove libusb.h libusb_strerror() implementation
      usbhid-dump: remove lib directory
      usbhid-dump: move .h files into src/ directory
      Makefile.am: add usbhid-dump.8 to distclean list
      usbhid-dump: some autoconf cleanup
      usbhid-dump: remove some dev_list functions that were never used
      dump_audiostreaming_interface(): remove unused variable
      usbmisc: initialize string buffer before reading from device.
      lsusb.py: drop trailing space on non-hub devices
      lsusb.py: strip whitespace from device strings

Jakub Wilk (2):
      lsusb(8): fix formatting
      lsusb(8): document --tree

Pino Toscano (1):
      lsusb.py: remove private paths for usb.ids

Rob Gill (1):
      Additional device classes for usb-devices script

Rosen Penev (1):
      usbhid-dump: Do not use rindex

Thomas Hebb (4):

usbreset.c
// SPDX-License-Identifier: GPL-2.0-only
/* usbreset -- send a USB port reset to a USB device */
/* Copyright (c) 2009-2016 Alan Stern */
/* To build:  gcc -o usbreset usbreset.c */

#include <stdint.h>
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <string.h>
#include <ctype.h>
#include <limits.h>
#include <dirent.h>
#include <sys/ioctl.h>

#include <linux/usbdevice_fs.h>


/* static char *usbfs = NULL; */

struct usbentry {
	int bus_num;
	int dev_num;
	int vendor_id;
	int product_id;
	char vendor_name[128];
	char product_name[128];
};

static char *sysfs_attr(const char *dev, const char *attr)
{
	int fd, len = 0;
	char path[PATH_MAX];
	static char buf[129];

	memset(buf, 0, sizeof(buf));
	snprintf(path, sizeof(path) - 1, "/sys/bus/usb/devices/%s/%s", dev, attr);

	fd = open(path, O_RDONLY);
	if (fd >= 0) {
		len = read(fd, buf, sizeof(buf) - 1);
		close(fd);
	}

	while (--len > 0 && isspace(buf[len]))
		buf[len] = 0;

	return (len >= 0) ? buf : NULL;
}

static struct usbentry *parse_devlist(DIR *d)
{
	char *attr;
	struct dirent *e;
	static struct usbentry dev;

	do {
		e = readdir(d);

		if (!e)
			return NULL;
	} while (!isdigit(e->d_name[0]) || strchr(e->d_name, ':'));

	memset(&dev, 0, sizeof(dev));
	dev.vendor_id = -1;
	dev.product_id = -1;

	attr = sysfs_attr(e->d_name, "busnum");
	if (attr)
		dev.bus_num = strtoul(attr, NULL, 10);

	attr = sysfs_attr(e->d_name, "devnum");
	if (attr)
		dev.dev_num = strtoul(attr, NULL, 10);

	attr = sysfs_attr(e->d_name, "idVendor");
	if (attr)
		dev.vendor_id = strtoul(attr, NULL, 16);

	attr = sysfs_attr(e->d_name, "idProduct");
	if (attr)
		dev.product_id = strtoul(attr, NULL, 16);

	attr = sysfs_attr(e->d_name, "manufacturer");
	if (attr)
		strcpy(dev.vendor_name, attr);

	attr = sysfs_attr(e->d_name, "product");
	if (attr)
		strcpy(dev.product_name, attr);

	if (dev.bus_num && dev.dev_num && dev.vendor_id >= 0 && dev.product_id >= 0)
		return &dev;

	return NULL;
}

static void list_devices(void)
{
	DIR *devs = opendir("/sys/bus/usb/devices");
	struct usbentry *dev;

	if (!devs)
		return;

	while ((dev = parse_devlist(devs)) != NULL)
		printf("  Number %03d/%03d  ID %04x:%04x  %s\n",
			   dev->bus_num, dev->dev_num,
			   dev->vendor_id, dev->product_id,
			   dev->product_name);

	closedir(devs);
}

static struct usbentry *find_device(int *bus, int *dev, int *vid, int *pid,
				    const char *product)
{
	DIR *devs = opendir("/sys/bus/usb/devices");

	struct usbentry *e, *match = NULL;

	if (!devs)
		return NULL;

	while ((e = parse_devlist(devs)) != NULL)
		if ((bus && (e->bus_num == *bus) && (e->dev_num == *dev)) ||
			(vid && (e->vendor_id == *vid) && (e->product_id == *pid)) ||
			(product && !strcasecmp(e->product_name, product))) {
			match = e;
			break;
		}

	closedir(devs);

	return match;
}

static void reset_device(struct usbentry *dev)
{
	int fd;
	char path[PATH_MAX];

	snprintf(path, sizeof(path) - 1, "/dev/bus/usb/%03d/%03d",
		dev->bus_num, dev->dev_num);

	printf("Resetting %s ... ", dev->product_name);

	fd = open(path, O_WRONLY);
	if (fd  > -1) {
		if (ioctl(fd, USBDEVFS_RESET, 0) < 0)
			printf("failed [%s]\n", strerror(errno));
		else
			printf("ok\n");

		close(fd);
	} else {
		printf("can't open [%s]\n", strerror(errno));
	}
}


int main(int argc, char **argv)
{
	int id1, id2;
	struct usbentry *dev;

	if ((argc == 2) && (sscanf(argv[1], "%3d/%3d", &id1, &id2) == 2))
		dev = find_device(&id1, &id2, NULL, NULL, NULL);
	else if ((argc == 2) && (sscanf(argv[1], "%4x:%4x", &id1, &id2) == 2))
		dev = find_device(NULL, NULL, &id1, &id2, NULL);
	else if ((argc == 2) && strlen(argv[1]) < 128)
		dev = find_device(NULL, NULL, NULL, NULL, argv[1]);
	else {
		printf("Usage:\n"
		       "  usbreset PPPP:VVVV - reset by product and vendor id\n"
		       "  usbreset BBB/DDD   - reset by bus and device number\n"
		       "  usbreset \"Product\" - reset by product name\n\n"
		       "Devices:\n");
		list_devices();
		return 1;
	}

	if (!dev) {
		fprintf(stderr, "No such device found\n");
		return 1;
	}

	reset_device(dev);
	return 0;
}
